packages/server/src/transformer.ts

```diff
@@ -86,9 +89,113 @@ export type DefaultDataTransformer = CombinedDataTransformer & {
  */
 export const defaultTransformer: DefaultDataTransformer = {
   _default: true,
   input: { serialize: (obj) => obj, deserialize: (obj) => obj },
   output: { serialize: (obj) => obj, deserialize: (obj) => obj },
 };
+
+export type TrpcTuplesonConfig =
+  | TsonAsyncOptions
+  | {
+      input: TsonAsyncOptions;
+      output: TsonAsyncOptions;
+    };
+
+/**
+ * @internal
+ */
+export function getDataTransformer<
+  TOpts extends {
+    transformer?: DataTransformerOptions | undefined;
+    experimental_tuplesonOptions?: TrpcTuplesonConfig;
+  },
+>(opts: TOpts | undefined) {
+  const transformer = opts?.experimental_tuplesonOptions
+    ? getTsonTransformer(opts)
+    : getLegacyTransformer(opts);
+
+  return transformer;
+}
+
+function getLegacyTransformer<TTransformer extends DataTransformerOptions>(
+  opts: { transformer?: TTransformer } | undefined,
+): CombinedDataTransformer {
+  const transformer = opts?.transformer;
+  if (!transformer) {
+    return defaultTransformer;
+  }
+  if ('input' in transformer) {
+    return transformer;
+  }
+  return {
+    input: transformer,
+    output: transformer,
+  };
+}
+
+function getTsonTransformer<
+  TOpts extends {
+    experimental_tuplesonOptions?: TrpcTuplesonConfig | undefined;
+    transformer?: DataTransformerOptions | undefined;
+  },
+>(opts: TOpts | undefined) {
+  if (!opts?.experimental_tuplesonOptions) {
+    throw new Error(
+      'TSON is not enabled. This should not happen; it is a bug.',
+    );
+  }
+
+  const officialTsonSyncHandlers = [
+    tsonBigint,
+    tsonDate,
+    tsonRegExp,
+    tsonSet,
+    tsonMap,
+    tsonUndefined,
+    tsonNumberGuard,
+    tsonURL,
+    tsonSymbol,
+  ] as const;
+
+  const tsonConfig = opts.experimental_tuplesonOptions;
+  const noLegacyTransformer = opts.transformer === undefined;
+
+  type Handler = TsonAsyncOptions['types'][number];
+
+  const old = getLegacyTransformer(opts);
+
+  function _createTson(inOut: 'input' | 'output') {
+    const userSyncHandlers = (
+      'input' in tsonConfig ? tsonConfig[inOut] : tsonConfig
+    ).types?.filter(
+      (type): type is Exclude<Handler, { async: true }> => !type.async,
+    );
+
+    return createTson({
+      ...tsonConfig,
+      types: [
+        ...(userSyncHandlers ?? officialTsonSyncHandlers),
+        ...(noLegacyTransformer
+          ? // enforced to mitigate prototype pollution attacks
+            // unfortunately not compatible with the legacy transformer API
+            [tsonUnknownObjectGuard]
+          : [
+              {
+                key: 'unknown',
+                test: () => true,
+                serialize: (data: unknown) => old[inOut].serialize(data),
+                deserialize: (data: unknown) => old[inOut].deserialize(data),
+              },
+            ]),
+      ],
+    });
+  }
+
+  return {
+    input: _createTson('input'),
+    output: _createTson('output'),
+  };
```

Comment:
Won't creating the transformer here _and_ in the resolveHttp create two different instances that have different nonce and contains two different "heads"?

Will likely not cause any issues, but it'd be nicer if they'd use the same instance

Reply:
Hmm, did I miss some lazy initialization of the revolvers? I didn't realize I was instantiating twice. Can you link the line?